packages/angular_devkit/build_angular/src/webpack/plugins/karma/karma.ts

```diff
@@ -169,29 +169,42 @@ const init: any = (config: any, emitter: any) => {
   compiler.hooks.watchRun.tapAsync('karma', (_: any, callback: () => void) => handler(callback));
   compiler.hooks.run.tapAsync('karma', (_: any, callback: () => void) => handler(callback));
 
+  webpackMiddleware = webpackDevMiddleware(compiler, webpackMiddlewareConfig);
+  emitter.on('exit', (done: any) => {
+    webpackMiddleware.close();
+    done();
+  });
+
   function unblock() {
     isBlocked = false;
     blocked.forEach((cb) => cb());
     blocked = [];
   }
 
   let lastCompilationHash: string | undefined;
-  compiler.hooks.done.tap('karma', (stats) => {
-    if (stats.hasErrors()) {
-      lastCompilationHash = undefined;
-    } else if (stats.hash != lastCompilationHash) {
-      // Refresh karma only when there are no webpack errors, and if the compilation changed.
-      lastCompilationHash = stats.hash;
-      emitter.refreshFiles();
-    }
-    unblock();
-  });
+  let isFirstRun = true;
+
+  return new Promise<void>((resolve) => {
+    compiler.hooks.done.tap('karma', (stats) => {
```

Comment:
What if Karma fails? Will this callback still be executed or do we need a separate error handler to reject the `Promise`?

Reply:
I think you meant the webpack fails, yeah this will always be executed.